Replace EC Private Key with PKCS#8 PEMs

alyyousuf7 opened the pull request:
This is to replace EC Private Keys with PKCS#8 in order to eliminate usage of MD5 in swarmkit when the keys are encrypted or decrypted.

aaronlehmann commented:
ping @diogomonica @cyli

Not my area of expertise, but one thought that comes to mind is that we need to be very careful about preserving backward compatibility with the old keys. Let's make sure we have tests that involve the old type of key, if we don't have those already.

cyli commented:
We currently don't have any test fixtures with the old type of key, since it includes the decryption key headers and we'd probably want to include the raft store along with that.  But that's a good idea in case we ever want to change the encryption algo/header format as well.  We should probably have some sample old data directories and always make sure we can read from disk.

diogomonica commented:
Agreed on adding old-style key data directories to make sure we can load them with the new code.

alyyousuf7 commented:
@aaronlehmann @cyli 
Are you referring to PKCS#1 when you say "old type of key"?

If we support PKCS#1 for backward compatibility, we will be using MD5 which will go against FIPS compliance.

aaronlehmann commented:
Keeping backward compatibility seems more important to me than removing MD5 from the code completely.

diogomonica commented:
@alyyousuf7 this code has to be backwards compatible for now, so we have to be able to load old style keys for a few versions before completely removing it.

endophage commented:
@diogomonica @aaronlehmann would it be acceptable for the backwards compatibility to only be operational in non-FIPS mode (i.e. the default configuration)? Having MD5 available at all when FIPS mode is enabled creates some serious headaches.

diogomonica commented:
@endophage yeah, I think it's fine if "when in FIPS mode" the only supported thing is PKCS#8

aaronlehmann commented:
Sorry if it's a stupid question... what does it mean to be "in FIPS mode"?

Does that mean we'd add a flag to disable legacy algorithms for compliance reasons?

diogomonica commented:
@aaronlehmann yes.

endophage commented:
@aaronlehmann we've been doing it with an env variable in other places with the idea being all apps can share the same env var. By enabling it once, everything in the docker stack goes in to FIPS mode.

aaronlehmann commented on keyutils/keyutils.go:
```diff
@@ -0,0 +1,87 @@
+package keyutils
```
Let's move this package under `ca`.

aaronlehmann commented on pkcs8/pkcs8.go:
```diff
@@ -0,0 +1,247 @@
+package pkcs8
```
Let's move this package under `ca`.

aaronlehmann commented on keyutils/keyutils.go:
```diff
@@ -0,0 +1,87 @@
+	if isPKCS8(block.Bytes) {
+		return pkcs8.IsEncryptedPEMBlock(block)
+	} else if FIPSEnabled() {
+		panic("Unsupported key format")
```
I don't think this panic is the right thing to do if we encounter an unsupported key format. The function should probably return an error in this case.

aaronlehmann commented on keyutils/keyutils.go:
```diff
@@ -0,0 +1,87 @@
+const FIPSEnvVar = "GOFIPS"
+
+// FIPSEnabled returns true if FIPS mode is enabled
+func FIPSEnabled() bool {
```
It looks like `FIPSEnabled` is only used in this package, so it does not need to be exported.

aaronlehmann commented on keyutils/keyutils_test.go:
```diff
@@ -0,0 +1,191 @@
+	assert.True(t, FIPSEnabled())
+	// Enable FIPS mode
+	os.Setenv(FIPSEnvVar, "1")
+	defer os.Unsetenv(FIPSEnvVar)
```
This looks odd. The environment variable is set and unset, with nothing in between.

aaronlehmann commented on ca/certificates.go:
```diff
@@ -815,6 +817,11 @@ func CreateRootCA(rootCN string) (RootCA, error) {
 		return RootCA{}, err
 	}
 
+	key, err = pkcs8.ConvertECPrivateKeyPEM(key)
```
By converting the key to pkcs8, won't this CA key only be usable by nodes running a recent enough version?

It seems better to me to only do this when `GOFIPS` is set.

alyyousuf7 replied:
It looks like it won't work if the CA key is encrypted using the deprecated `SWARM_ROOT_CA_PASSPHRASE` environment variable.

cyli commented on ca/pkcs8/pkcs8.go:
```diff
@@ -0,0 +1,247 @@
+		mode := cipher.NewCBCDecrypter(block, iv)
+		mode.CryptBlocks(encryptedKey, encryptedKey)
+
+		// Remove padding
```
Just for my own understanding, when comparing this to https://github.com/docker/notary/pull/1130/files#diff-14d2480b5aeb5325dbdc2c70ae8fd052R215, and to https://github.com/youmark/pkcs8/blob/master/pkcs8.go#L143, why does this implementation need to remove padding?

I assume for `notary`, perhaps the subsequence `asn1.Unmarshal` after the `mode.CryptBlocks` call ignores the extra padding.  It seems like [`x509.ParsePKCS8PrivateKey`](https://golang.org/src/crypto/x509/pkcs8.go?s=580:646#L16) also does an unmarshal immediately, though, and [`derhelpers.ParsePrivateKeyDER`](https://github.com/cloudflare/cfssl/blob/master/helpers/derhelpers/derhelpers.go#L16) calls `x509.ParsePKCS8PrivateKey` immediately.

alyyousuf7 replied:
When encoding the key to pem, it was adding an extra base64 chunk due to the padding which was unnecessary.

cyli replied:
Ah ok, does this same change need to be made to https://github.com/docker/notary/pull/1130/files#diff-14d2480b5aeb5325dbdc2c70ae8fd052R215?  The encrypting key -> PEM part seems the same.

cyli commented on ca/pkcs8/pkcs8_test.go:
```diff
@@ -0,0 +1,95 @@
+
+	decryptedDer, err := DecryptPEMBlock(block, []byte("ponies"))
+	require.NoError(t, err)
+	require.Equal(t, expectedBlock.Bytes, decryptedDer)
```
Thanks for adding!

cyli commented on ca/keyutils/keyutils_test.go:
```diff
@@ -0,0 +1,187 @@
+	require.Error(t, err)
+
+	_, err = ParsePrivateKeyPEMWithPassword([]byte(encryptedPKCS8), []byte("ponies"))
+	require.NoError(t, err)
```
Thanks for adding!

alyyousuf7 commented on ca/keyutils/keyutils.go:
```diff
@@ -0,0 +1,80 @@
+package keyutils
```
It actually supports both keys, as it uses `cfssl/helpers` internally to parse decrypted PEMs

cyli replied:
https://github.com/docker/swarmkit/pull/2246/files#diff-4d81ff593f40aff60e869a25f173c15dR78 specifically hardcodes EC private keys.  

Other than that, it's more just that we don't implement converting RSA keys.

diogomonica commented on ca/pkcs8/pkcs8.go:
```diff
@@ -0,0 +1,273 @@
+
+	// Use AES256-CBC mode, pad plaintext with PKCS5 padding scheme
+	padding := aes.BlockSize - len(pkey)%aes.BlockSize
+	if padding > 0 {
```
This code seems correct, but operating on the final key buffer directly when padding is asking for trouble.

- Let's define `n := len(pkey)` outside of the `if`
- Let's create the padding array first, fill it with `0..len(padding)` and only then we append the padding array to `pkey`

diogomonica commented on ca/pkcs8/pkcs8.go:
```diff
@@ -0,0 +1,273 @@
+		keyLen := len(encryptedKey)
+		padLen := int(encryptedKey[keyLen-1])
+		if padLen > keyLen || padLen > aes.BlockSize {
+			return nil, errors.New("pkcs8: invalid padding size")
```
Would be great to add an `invalid padding` test

diogomonica commented on integration/cluster.go:
```diff
@@ -409,3 +413,61 @@ func (c *testCluster) RotateRootCA(cert, key []byte) error {
 	}, opsTimeout)
 }
+
+func (c *testCluster) RotateUnlockKey() error {
```
This looks like this doesn't belong here. Failed rebase?

diogomonica replied:
Seems like this is just a refactor of the current tests.

alyyousuf7 replied:
This was to add tests to check if manager could lock/rotate/unlock in FIPS mode.

cyli commented:
From discussions with @aluzzardi and @friism, this is ok to merge provided that we document how to downgrade (this change would only prevent downgrading on managers that were autolocked, if the encrypted key is PKCS8 - the user can disable autolock, wait for the key to be decrypted on all manager nodes, and then downgrade), and if that fails, to provide a downgrade tool that can convert from an encrypted PKCS8 key to an encrypted PKCS1 key.

alyyousuf7 commented:
> we document how to downgrade

Where would we be documenting it exactly?

> to provide a downgrade tool that can convert from an encrypted PKCS8 key to an encrypted PKCS1 key.

Do you mean a separate CLI to which a user can provide an unlock key, the tool will then downgrade the encrypted PKCS8 keys to encrypted PKCS1 keys. Or do you have something else in mind?

cyli commented:
> Where would we be documenting it exactly?

I was going to open an issue in https://github.com/docker/docker.github.io - I'm not entirely sure which doc it should go in, but I'll include details in the issue of the instructions so they can decide.

> Do you mean a separate CLI to which a user can provide an unlock key, the tool will then downgrade the encrypted PKCS8 keys to encrypted PKCS1 keys.

Yep, that's it.  We have something similar in `cmd/swarm-rafttool` that decrypts the raft logs.

alyyousuf7 commented:
> Yep, that's it. We have something similar in cmd/swarm-rafttool that decrypts the raft logs.

Just for clarification, this is to change key format (to be compatible with old versions) *without disabling autolock*.

If disabling autolock was not an issue, we could make the keys compatible with old version of swarmkit simply by disabling autolock using 'swarmctl', which would result in unencrypted PKCS8 keys.

We want to be able to downgrade without disabling autolock.
Please correct me if I'm wrong :)

cyli commented:
@alyyousuf7 I believe that is correct - the documentation will say to disable autolock first and if desired, to re-enable later, but this is more a tool in case something goes wrong or if customers didn't want to disable autolock.  ccing @friism to verify that this is still a requirement.

alyyousuf7 commented:
I have created a CLI which will downgrade the manager key to PKCS1. Let me know if moving this to swarmctl would be a better idea.
@cyli @aluzzardi @friism

cyli commented:
@alyyousuf7 Apologies for the lack of response!  Probably `swarmctl` or `swarm-rafttool` would be a better place to put this.  `swarm-rafttool` already has some functionality for decrypting the raft logs.  In particular, `swarm-downgrade` may imply that it can downgrade the entire swarm, whereas it's just converting the key to a different format.

alyyousuf7 commented:
> `swarm-downgrade` may imply that it can downgrade the entire swarm

Agreed, moved it as a new command `downgrade-key` in `swarm-rafttool` and rebased the PR.

cyli commented on node/node.go:
```diff
@@ -971,6 +971,15 @@ func (n *Node) superviseManager(ctx context.Context, securityConfig *ca.Security
 
+// Downgrade reverts the node key to older format so that it can be
+// run on older version of swarmkit
+func (n *Node) Downgrade() error {
```
Apologies for the nitpick, similar to the comment about the command line tool being "downgrade-key" instead of just "downgrade", since this is a function on a node itself, would it make sense for this function to also be called "DowngradeKey"?

alyyousuf7 replied:
Resolved! :)

dperny commented:
LGTM!
